State machine designed

ArmanAghighi opened the pull request:
Added state machine and some state like Start , Game play , Loose , Win , pause 
@URAmiRBin

ArmanAghighi committed:
State machine designed

Added state machine and some state like Start , Game play , Loose , Win , pause

URAmiRBin commented on Assets/StateMachine/New Model.meta:
put the scripts into a folder called scripts,
`new model` is not a good name for this folder it does not indicate what is in there

URAmiRBin commented on Assets/StateMachine/New Model/BaseState.cs:
its just `State` the name `BaseState` contains redundant information

URAmiRBin commented on Assets/StateMachine/GameManager.cs:
```diff
@@ -0,0 +1,19 @@
+
+public class GameManager : MonoBehaviour
+{
+    StateManager _state;
```
`StateManager _state` is kinda misleading, is it a state or a state manager?

URAmiRBin commented on Assets/StateMachine/New Model/BaseState.cs:
```diff
@@ -0,0 +1,8 @@
+
+public abstract class BaseState
+{
+    public abstract void EnterState(StateManager state);
```
why `StateManager` should be passed to a state in order to enter that state?

URAmiRBin commented on Assets/StateMachine/New Model/OnGamePlayState.cs:
```diff
@@ -0,0 +1,30 @@
+    {
+        // Play sound
+        Debug.Log("Enter Game Play State");
+    }
+    public override void UpdateState(StateManager state)
```
spaces please

URAmiRBin commented on Assets/StateMachine/New Model/OnLoseState.cs:
find a way to make the states general, its a bad code smell when you need to create a class for each state, remember oop principles? classes are different from objects, in my opinion one state is just an object

URAmiRBin commented on Assets/StateMachine/New Model/StateManager.cs:
```diff
@@ -0,0 +1,36 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+public class StateManager : MonoBehaviour
```
its called a state machine

URAmiRBin replied:
also state machine should not be a monobehaviour

URAmiRBin commented on Assets/StateMachine/New Model/StateManager.cs:
```diff
@@ -0,0 +1,36 @@
+public class StateManager : MonoBehaviour
+{
+    private BaseState CurrentState;
+    public OnStartState StartState = new OnStartState();
+    public OnGamePlayState GamePlay = new OnGamePlayState();
+    public OnPauseState PauseState = new OnPauseState();
+    public OnWinState WinState = new OnWinState();
+    public OnLoseState LoseState = new OnLoseState();
```
we do not need these much classes, just initialize all the `State` s in awake and define their behaviour

URAmiRBin commented on Assets/StateMachine/StateMachineScene.unity.meta:
i expect the files to be added in their own commits, one commit and 20 files is a bad mess

ArmanAghighi committed:
changed requirement done

- changed the name of stateManager to stateMachine
- add some space between the methods

ArmanAghighi committed:
State machine optimization

use graph , event and action delegate to create an optimized state machine

ArmanAghighi committed:
Delete SceneTemplateSettings.json

URAmiRBin commented on Assets/StateMachine/Scripts/GameManager.cs:
```diff
@@ -0,0 +1,192 @@
+
+public class GameManager : MonoBehaviour
+{
+    private State stateEvent;
```
remove unused variables

URAmiRBin commented on Assets/StateMachine/Scripts/GameManager.cs:
```diff
@@ -0,0 +1,192 @@
+    private State winState;
+    private State loseState;
+    private State pauseState;
+    private State exitState;
```
```suggestion
    private State sleepState, startState, gameplayState, winState, loseState, pauseState, exitState;
```

URAmiRBin commented on Assets/StateMachine/Scripts/GameManager.cs:
```diff
@@ -0,0 +1,192 @@
+    private State stateEvent;
+    private StateMachine stateMachine;
+    private bool isStarted = false;
+    private State sleepState;
```
what is sleep state?

URAmiRBin commented on Assets/StateMachine/Scripts/GameManager.cs:
```diff
@@ -0,0 +1,192 @@
+    private State exitState;
+    void Start()
+    {
+        sleepState = new State(EnterSleepState, UpdateSleepState, ExitSleepState);
+        startState = new State(EnterStartState, UpdateStartState, ExitStartState);
+        gameplayState = new State(EnterGamePlayState, UpdateGamePlayState, ExitGamePlayState);
+        winState = new State(EnterWinState, UpdateWinState, ExitWinState);
+        loseState = new State(EnterLoseState, UpdateLoseState, ExitLoseState);
+        pauseState = new State(EnterPauseState, UpdatePauseState, ExitPauseState);
+        exitState = new State(EnterQuitState, UpdateQuitState, ExitQuitState);
```
wrap these in `CreateStates` function

URAmiRBin commented on Assets/StateMachine/Scripts/GameManager.cs:
```diff
@@ -0,0 +1,192 @@
+        
+        var gameLoopStateMachineGraph = new Graph<State>();
+        
+        gameLoopStateMachineGraph.AddVertex(sleepState);
+        gameLoopStateMachineGraph.AddVertex(startState);
+        gameLoopStateMachineGraph.AddVertex(gameplayState);
+        gameLoopStateMachineGraph.AddVertex(winState);
+        gameLoopStateMachineGraph.AddVertex(loseState);
+        gameLoopStateMachineGraph.AddVertex(pauseState);
+        gameLoopStateMachineGraph.AddVertex(exitState);
+        
+        gameLoopStateMachineGraph.AddEdge(sleepState,startState);
+        gameLoopStateMachineGraph.AddEdge(startState, gameplayState);
+        gameLoopStateMachineGraph.AddEdge(startState,pauseState);
+        gameLoopStateMachineGraph.AddEdge(gameplayState,winState);
+        gameLoopStateMachineGraph.AddEdge(gameplayState,loseState);
+        gameLoopStateMachineGraph.AddEdge(gameplayState,pauseState);
+        gameLoopStateMachineGraph.AddEdge(pauseState,startState);
+        gameLoopStateMachineGraph.AddEdge(pauseState,gameplayState);
+        gameLoopStateMachineGraph.AddEdge(pauseState,exitState);
+        gameLoopStateMachineGraph.AddEdge(winState,exitState);
+        gameLoopStateMachineGraph.AddEdge(loseState,exitState);
```
wrap these in `CreateStateMachineGraph` function

URAmiRBin commented on Assets/StateMachine/Scripts/GameManager.cs:
```diff
@@ -0,0 +1,192 @@
+    {
+        stateMachine.Update();
+        if (Input.GetKeyDown(KeyCode.S))
+        {//is in sleep state
```
no comments in code

URAmiRBin commented on Assets/StateMachine/Scripts/GameManager.cs:
```diff
@@ -0,0 +1,192 @@
+        stateMachine.Update();
+        if (Input.GetKeyDown(KeyCode.S))
+        {//is in sleep state
+            if (!isStarted)
+            {            
+                stateMachine.Transition(startState);
+                isStarted = true;
+            }
+            else
+            {
+                Debug.Log("Can Not change the state back to start");
+                return;
+            }
+        }
+        if(isStarted && Input.GetKeyDown(KeyCode.G))
+            stateMachine.Transition(gameplayState);//start gameplay state
+        if (!isStarted) 
+        {// before start state
+            if (Input.GetKeyDown(KeyCode.L) || Input.GetKeyDown(KeyCode.W))
+                Debug.Log("Can Not change the state to win or lose");
+            if(Input.GetKeyDown(KeyCode.P)) 
+                stateMachine.Transition(pauseState);
```
optimize this code, cant be read well

URAmiRBin commented on Assets/StateMachine/Scripts/GameManager.cs:
```diff
@@ -0,0 +1,192 @@
+                stateMachine.Transition(pauseState);
+        }
+        else 
+        {// is in Game Play state
```
no comment needed

URAmiRBin commented on ProjectSettings/ProjectSettings.asset:
this file should not be included in your branch

ArmanAghighi committed:
Change requirements are done

URAmiRBin commented on Assets/StateMachine/Scripts/GameManager.cs:
```diff
@@ -0,0 +1,173 @@
+    {
+        void CreateState()
+        {
+            standbyState = new State(EnterstandbyState, UpdatestandbyState, ExitstandbyState);
+            startState = new State(EnterStartState, UpdateStartState, ExitStartState);
+            gameplayState = new State(EnterGamePlayState, UpdateGamePlayState, ExitGamePlayState);
+            winState = new State(EnterWinState, UpdateWinState, ExitWinState);
+            loseState = new State(EnterLoseState, UpdateLoseState, ExitLoseState);
+            pauseState = new State(EnterPauseState, UpdatePauseState, ExitPauseState);
+            exitState = new State(EnterQuitState, UpdateQuitState, ExitQuitState);
+        }
+
+        var gameLoopStateMachineGraph = new Graph<State>();
+        
+        void CreateStateMachineGraph()
+        {
+            gameLoopStateMachineGraph.AddVertex(standbyState);
+            gameLoopStateMachineGraph.AddVertex(startState);
+            gameLoopStateMachineGraph.AddVertex(gameplayState);
+            gameLoopStateMachineGraph.AddVertex(winState);
+            gameLoopStateMachineGraph.AddVertex(loseState);
+            gameLoopStateMachineGraph.AddVertex(pauseState);
+            gameLoopStateMachineGraph.AddVertex(exitState);
+        
+            gameLoopStateMachineGraph.AddEdge(standbyState,startState);
+            gameLoopStateMachineGraph.AddEdge(startState, gameplayState);
+            gameLoopStateMachineGraph.AddEdge(startState,pauseState);
+            gameLoopStateMachineGraph.AddEdge(gameplayState,winState);
+            gameLoopStateMachineGraph.AddEdge(gameplayState,loseState);
+            gameLoopStateMachineGraph.AddEdge(gameplayState,pauseState);
+            gameLoopStateMachineGraph.AddEdge(pauseState,startState);
+            gameLoopStateMachineGraph.AddEdge(pauseState,gameplayState);
+            gameLoopStateMachineGraph.AddEdge(pauseState,exitState);
+            gameLoopStateMachineGraph.AddEdge(winState,exitState);
+            gameLoopStateMachineGraph.AddEdge(loseState,exitState);
+        }
+        CreateState();
+        CreateStateMachineGraph();
+        stateMachine = new StateMachine(gameLoopStateMachineGraph, standbyState);
+    }
```
```suggestion
    void Start()
    {
        var gameLoopStateMachineGraph = new Graph<State>();
        CreateState();
        CreateStateMachineGraph();
        stateMachine = new StateMachine(gameLoopStateMachineGraph, standbyState);

            
        void CreateState()
        {
            standbyState = new State(EnterstandbyState, UpdatestandbyState, ExitstandbyState);
            startState = new State(EnterStartState, UpdateStartState, ExitStartState);
            gameplayState = new State(EnterGamePlayState, UpdateGamePlayState, ExitGamePlayState);
            winState = new State(EnterWinState, UpdateWinState, ExitWinState);
            loseState = new State(EnterLoseState, UpdateLoseState, ExitLoseState);
            pauseState = new State(EnterPauseState, UpdatePauseState, ExitPauseState);
            exitState = new State(EnterQuitState, UpdateQuitState, ExitQuitState);
        }
        
        void CreateStateMachineGraph()
        {
            gameLoopStateMachineGraph.AddVertex(standbyState);
            gameLoopStateMachineGraph.AddVertex(startState);
            gameLoopStateMachineGraph.AddVertex(gameplayState);
            gameLoopStateMachineGraph.AddVertex(winState);
            gameLoopStateMachineGraph.AddVertex(loseState);
            gameLoopStateMachineGraph.AddVertex(pauseState);
            gameLoopStateMachineGraph.AddVertex(exitState);
        
            gameLoopStateMachineGraph.AddEdge(standbyState,startState);
            gameLoopStateMachineGraph.AddEdge(startState, gameplayState);
            gameLoopStateMachineGraph.AddEdge(startState,pauseState);
            gameLoopStateMachineGraph.AddEdge(gameplayState,winState);
            gameLoopStateMachineGraph.AddEdge(gameplayState,loseState);
            gameLoopStateMachineGraph.AddEdge(gameplayState,pauseState);
            gameLoopStateMachineGraph.AddEdge(pauseState,startState);
            gameLoopStateMachineGraph.AddEdge(pauseState,gameplayState);
            gameLoopStateMachineGraph.AddEdge(pauseState,exitState);
            gameLoopStateMachineGraph.AddEdge(winState,exitState);
            gameLoopStateMachineGraph.AddEdge(loseState,exitState);
        }
    }
```

URAmiRBin committed:
Reorder functions
